Private Access Token protocol rework (#63)
tfpauly/privacy-proxy: draft-private-access-tokens.md

identity and actions, and each entity only knows enough to serve its
function. The pieces of information are identified in {{terms}}.

> The Mediator knows the Client's identity , the Issuer being targeted

Reviewer: nit: "identity," (extra space)

Author: ```suggestion
The Mediator knows the Client's identity, the Issuer being targeted
```

is held by the Issuer.

ISSUER_NAME_KEY:
: The public key used to encrypt the ORIGIN_NAME in request from Clients
to the Issuer, so that Mediators cannot learn the ORIGIN_NAME value. Each
ISSUER_NAME_KEY is used across all requests on the Issuer, for different Origins.

ORIGIN_NAME:
: The name of the Origin that requests and verifies Private Access Tokens.

ANON_ORIGIN_ID:
: An identifier that is generated by the Client and marked on requests to the
Mediator, which represents a specific Origin anonymously. The Client generates
a stable ANON_ORIGIN_ID for each ORIGIN_NAME, to allow the Mediator to count
token access without learning the ORIGIN_NAME.

CLIENT_GROUP_ID:
: An identifier chosen by the Mediator and sent to the Issuer, which represents
a random cohort of Clients. Each Client is assigned a new CLIENT_GROUP_ID for each
Issuer it uses every time a new policy window begins.

GROUP_ORIGIN_ID:
: An identifier chosen by the Issuer and sent to the Mediator, which corresponds
to the tuple of an ORIGIN_NAME and a CLIENT_GROUP_ID. This identifier is used by
the Mediator to ensure that Clients cannot create a mismatch between an

> ORIGIN_NAME and an ANON_ORIGIN_ID.

Reviewer 2: The purpose of this should be to allow the Mediator and the Issuer to communicate about the Origin, including for the Mediator to enforce the Issuer's policy, right?

Author: It's just for the issuer to communicate to the mediator something about the origin that lets the mediator detect client's trying to abuse the system. The actual per-origin limit is enforce by the issuer.

# API Endpoints {#setup}

Issuers MUST provide the following information available via the corresponding
API endpoints:

- ISSUER_NAME_KEY: /.well-known/issuer-name-key
- ISSUER_POLICY_WINDOW: /.well-known/issuer-policy

The content of issuer name public key is a `KeyConfig` as defined in {{!OHTTP=I-D.thomson-http-oblivious}}
to use when encrypting the ORIGIN_NAME in issuance requests. The response uses media type
"application/ohttp-keys".

> The policy window (ISSUER_POLICY_WINDOW) is store as a resource of media type "application/json",

Reviewer 2: ```suggestion
The policy window (ISSUER_POLICY_WINDOW) is a resource of media type "application/json",
```

This can be stored at the Mediator and Issuer in any form; it's just communicated as app/json

for generating access tokens. For example, one template for the Issuer might
be https://issuer.net/access-token-request.

Mediators advertise a URI template {{!RFC6570}} for proxying protocol messages
to Issuers. For example, one template for the Mediator might

> be https://mediator.net/relay-access-token-request.

Reviewer 2: Why is this a template? The example is simply a URL.

## Token Challenge and Redemption {#scheme}

Token redemption is performed using HTTP Authentication ({{!RFC7235}}), with
the scheme "PrivateAccessToken". Origins challenge Clients to present a unique,

> single-use token from a specific Issuer. Once Clients have received a token
> from that Issuer, they present the token to the Origin.

Reviewer 2: ```suggestion
single-use token from a specific Issuer. Once a Client has received a token
from that Issuer, it presents the token to the Origin.
```

> ### Token Challenge

Reviewer 2: ```suggestion
### Token Challenge {#challenge}
```

- "redemption_nonce" is a fresh 32-byte nonce generated for each redemption request.

> When used in authentication challenges, the "PrivateAccessToken" scheme uses the

Reviewer 2: ```suggestion
When used in an authentication challenge, the "PrivateAccessToken" scheme uses the
```

WWW-Authenticate: PrivateAccessToken challenge=abc... token-key=123... name-key=456...
~~~

> Upon receipt of this challenge, Clients use the message and keys in the Issuance protocol
> as described in {{issuance}}. If the TokenChallenge has a version field the Client

Reviewer 2: ```suggestion
Upon receipt of this challenge, a Client uses the message and keys in the Issuance protocol
(see {{issuance}}). If the TokenChallenge has a version field the Client
```

does not recognize or support, it MUST NOT parse or respond to the challenge.
This document defines version 1, which indicates use of private tokens based on
RSA Blind Signatures {{BLINDSIG}}, and determines the rest of the structure contents.

### Token Redemption

> The output of the issuance protocol is a token that corresponds to the challenge.

Reviewer 2: ```suggestion
The output of the issuance protocol is a token that corresponds to the Origin's challenge (see {{challenge}}).
```

> The Token is a structure that begins with a single byte that indicates a version, which

Reviewer 2: ```suggestion
A token is a structure that begins with a single byte that indicates a version, which
```

## Issuance {#issuance}

Token issuance involves a Client, Mediator, and Issuer, with the following steps:

> 1. The Client sends a token request to the Mediator, using keys for a specific Issuer

Reviewer 2: ```suggestion
1. The Client sends a token request to the Mediator, encrypted using an Issuer-specific key
```

> 1. The Mediator validates the request and sends the request to the Issuer

Reviewer 2: ```suggestion
1. The Mediator validates the request and proxies the request to the Issuer
```

> 1. The Issuer decrypts the request and sends a response back via the Mediator

Reviewer 2: ```suggestion
1. The Issuer decrypts the request and sends a response back to the Mediator
```

> 1. The Mediator checks the response and sends the response to the Client

Reviewer 2: ```suggestion
1. The Mediator verifies the response and proxies the response to the Client
```

  corresponding to the Origin identified by TokenChallenge.origin_name.
- Issuer name public key (ISSUER_NAME_KEY), a public key used to encrypt requests
  corresponding to the Issuer identified by TokenChallenge.issuer_name.
  
Clients also need to be able to generate an ANON_ORIGIN_ID value that corresponds
to the ORIGIN_NAME, to send in requests to the Mediator.

ANON_ORIGIN_ID MUST be a stable and unpredictable 32-byte value computed by the Client.
Clients MUST NOT change this value across token requests for the same ORIGIN_NAME. Doing

> so will result in token issuance failure.

Reviewer 2: It's not clear to me how this results in token issuance failure.

One possible mechanism for implementing this identifier is for the Client to store a mapping
between the ORIGIN_NAME and a randomly generated ANON_ORIGIN_ID for future requests. Alternatively,
the Client can compute a PRF keyed by a per-client secret (CLIENT_SECRET) over the ORIGIN_NAME,
e.g., ANON_ORIGIN_ID = HKDF(secret=CLIENT_SECRET, salt="", info=ORIGIN_NAME).

### Mediator State {#mediator-state}

> Issuance assumes that Mediators maintain local state for each Client. The mechanism

Reviewer 2: ```suggestion
Issuance requires Mediators to maintain state for each Client. The mechanism
```

of identifying a Client is specific to each Mediator, and is not defined in this document.
As examples, the Mediator could use device-specific certificates or account authentication
to identify a Client.

Mediators are expected to know the ISSUER_POLICY_WINDOW for any ISSUER_NAME to which
they allow access. This information can be retrieved using the URIs defined in {{setup}}.

> For each pair of Client and Issuer, the Mediator maintains a specific policy window
> start and end time for each Issuer requested by the Client. Associated with this window,

Reviewer 2: ```suggestion
For each Client-Issuer pair, a Mediator maintains a policy window
start and end time for each Issuer from which a Client requests a token.  Associated with this window,
```

the Mediator determines a CLIENT_GROUP_ID value for the Client-Issuer pair, which

> changes every time the policy window elapses. The Mediator additionally maintains

Reviewer 2: ```suggestion
changes every time the policy window elapses. For each Client, the Mediator maintains
```

a counter of successful tokens issued within the policy window for each ANON_ORIGIN_ID
requested by the Client, as well as a GROUP_ORIGIN_ID that it received for each
ANON_ORIGIN_ID.

The Mediator assigns CLIENT_GROUP_ID values to Clients in way that segments them into

> groups of roughly equal size. For example, at any time, Clients could be spread

Reviewer 2: Say something about the group size -- that the Mediator chooses them, and the tradeoff is that an overly large group size makes it possible for a Mediator to learn about the Origin, and that an overly small group size makes it possible for the Issuer to learn about the Client. A Mediator SHOULD start by spreading users among as many groups as it can. With a large enough number of Clients, a Mediator SHOULD use at least ten groups at any given time, with each group having no fewer than thousands of Clients in it.

Reviewer: The incentives for choosing a group size are a bit odd, that it seems like the mediator should always want to minimzize it, while the issuer wants to maximize it. I guess issuers will have minimum requirements for mediators that partner with them and vice versa.

Reviewer 3: It's not clear to me how a large group reveals information about the origin. @janaiyengar, can you elaborate?

Author: @dvorak42 in order to learn info, I think it would actually be mediators preferring fewer larger groups (to learn about origins) and issuers preferring more smaller groups (to track clients)

Author: @janaiyengar I've taken your text and added it in here.

across 10 different CLIENT_GROUP_ID values. The specific values for the CLIENT_GROUP_ID
are integers determined by the Mediator, and is not reused across different
policy windows in a short period of time. One simple mechanism would be to have a counter
for groups, up to the maximum value of uint64, before wrapping around.

A specific CLIENT_GROUP_ID can be in one of three states:

- Open: Clients can be assigned this CLIENT_GROUP_ID upon first access to a given
Issuer for their specific policy window. This state lasts exactly the number
of seconds in ISSUER_POLICY_WINDOW.
- Draining: Clients can no longer be assigned this CLIENT_GROUP_ID for new access,
but some Clients are still using this CLIENT_GROUP_ID. This state also lasts exactly
the number of seconds in ISSUER_POLICY_WINDOW.
- Closed: No Clients are using this CLIENT_GROUP_ID, and no new Clients can have

> have it assigned.

Reviewer 2: When should this be reaped? Perhaps after a large enough period of time, say ten times the policy window period?

Any particular CLIENT_GROUP_ID MUST only be used for a maximum of twice the
ISSUER_POLICY_WINDOW, in seconds. This is the combination of the open and draining
states.

### Issuer State {#issuer-state}

Issuers are expected to have the private key that corresponds to ISSUER_NAME_KEY,
which allows them to decrypt the ORIGIN_NAME values in requests.

Issuers also need to know the current ORIGIN_TOKEN_KEY public key and corresponding
private key, for each ORIGIN_NAME that is served by the Issuer.

During issuance, Issuers help Mediators validate requests by provding a GROUP_ORIGIN_ID
in responses that is derived from a pair of ORIGIN_NAME and CLIENT_GROUP_ID values.

GROUP_ORIGIN_ID MUST be a stable and unpredictable 32-byte value computed by the Issuer.
Issuers MUST NOT change this value across token requests for the same ORIGIN_NAME and

> CLIENT_GROUP_ID.

Reviewer 2: ```suggestion
CLIENT_GROUP_ID within a policy window period.
```

This can change if the CLIENT_GROUP_ID has been reaped and is being reused after a while.

One possible mechanism for generating GROUP_ORIGIN_ID is for the Issuer to compute a PRF
keyed by a per-Issuer secret (issuer_secret) over the ORIGIN_NAME and CLIENT_GROUP_ID,
e.g., GROUP_ORIGIN_ID = HKDF(secret=issuer_secret, salt=CLIENT_GROUP_ID, info=ORIGIN_NAME).

Issuers also can keep state about previous CLIENT_GROUP_ID values they have seen from
a specific Mediator. Issuers SHOULD validate that a give CLIENT_GROUP_ID value is only

> used for at most twice the length of the ISSUER_POLICY_WINDOW. If a CLIENT_GROUP_ID

Reviewer 2: Over what period of time? Meaning, when does the Issuer stop tracking the CLIENT_GROUP_ID? This should be the same time as the Mediator is prohibited from reusing it -- "ten times the policy window".

Author: Clarified, and gave 10x policy window as the minimum

is used beyond that, Issuers SHOULD reject token requests. This prevents Mediators
from being able to correlate per-origin activity over time. Similarly, Issuers SHOULD
validate that a Mediator is sending multiple CLIENT_GROUP_ID values, and has a roughly

> equal split between requests that use various CLIENT_GROUP_ID values. If one

Reviewer: Is there is a difference between a mediator just having two groups or a single group per epoch? Or does the issuer need some way to indicate a minimum number of groups it expects to see?

Author: Yeah, there does need to be an effective minimum. I don't know if we need to have it communicated (it could go in issuer policy, with the policy window, if it does), or we should just have a minimum in the protocol as a MUST.

Reviewer 2: I've proposed above. It's best if we recommend a value here to start with, since Mediators will have no idea of how to determine this value.

Author: I've updated the text for the mediator requirements

If any of these conditions is not met, the Issuer MUST return an HTTP 400 error to the Mediator,
which will forward the error to the client.

If the request is valid, the Issuer then can use the value from "Client-Token-Count" to determine if
the Client is allowed to receive a token for this Origin during the current policy window. If the
Issuer refuses to issue more tokens, it responds with an HTTP 429 (Too Many Requests) error to the
Mediator, which will forward the error to the client.

The Issuer determines the correct ORIGIN_TOKEN_KEY by using the decrypted ORIGIN_NAME value. Issuers
are expected to be able to deterministically select the correct key based on information sent in

> the request. Clients do not indicate the ORIGIN_TOKEN_KEY to use, to prevent Origins from choosing

Reviewer: Does this get leaked by the choice of key that the ENCRYPTED_ORIGIN_NAME was encrypted to? (can't the Origin provide per-client ISSUER_NAME_KEYs?)

Author: The Mediator is responsible for checking the key ID of ISSUER_NAME_KEY to prevent this.

of tokens a given user will obtain.

Origins request tokens from Clients and, upon successful redemption, the Origin knows
the Client was able to request a token for the given (ANON_CLIENT_ID, ORIGIN_NAME) tuple. As a
the Client was able to request a token for the given ORIGIN_NAME tuple. As a
result, the Origin knows this is an authentic client.

> 

Reviewer: Might need to expand on the privacy/security implications in the considerations sections below on the mediator's choice of group size? It seems like group sizes of 1 or infinity (all the clients within an epoch) could have some slight concerns without the Issuer enforcing some size requirements?

Author: Yes, agreed.
